base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -880,85 +880,125 @@ where B: BlockchainBackend
-        // we can link it to the main chain. Full block validation only happens when the proof of work is higher than
-        // the main chain and we want to add the block to the main chain.
-        let block_add_result = add_block(
-            &mut *db,
-            &self.config,
-            &self.consensus_manager,
-            &*self.validators.block,
-            &*self.validators.header,
-            self.consensus_manager.chain_strength_comparer(),
-            candidate_block,
-        )?;
+            let before_lock = timer.elapsed();
+            let mut db = self.db_write_access()?;
+            let after_lock = timer.elapsed();
+
+            trace!(
+                target: LOG_TARGET,
+                "[add_block] acquired write access db lock for block #{} '{}' in {:.2?}",
+                new_height,
+                block_hash.to_hex(),
+                after_lock - before_lock,
+            );
+
+            // the only fast check we can perform that is slightly expensive to fake is a min difficulty check, this is
+            // done as soon as we receive the block before we do any processing on it. A proper proof of
+            // work is done as soon as we can link it to the main chain. Full block validation only happens
+            // when the proof of work is higher than the main chain and we want to add the block to the main
+            // chain.
+            let res = add_block(
+                &mut *db,
+                &self.config,
+                &self.consensus_manager,
+                &*self.validators.block,
+                &*self.validators.header,
+                self.consensus_manager.chain_strength_comparer(),
+                candidate_block,
+            )?;
+            debug!(
+                target: LOG_TARGET,
+                "[add_block] released write access db lock for block #{} in {:.2?}, `add_block` result: {}",
+                new_height, timer.elapsed() - after_lock, res
+            );
+            res
+        };
 
+        // If blocks were added and the node is in pruned mode, perform pruning
         if block_add_result.was_chain_modified() {
+            let metadata = {
+                let db = self.db_read_access()?;
+                db.fetch_chain_metadata()?
+            };
             info!(
                 target: LOG_TARGET,
                 "Best chain is now at height: {}",
-                db.fetch_chain_metadata()?.height_of_longest_chain()
+                metadata.height_of_longest_chain()
             );
-            // If blocks were added and the node is in pruned mode, perform pruning
-            prune_database_if_needed(&mut *db, self.config.pruning_horizon, self.config.pruning_interval)?;
-        }
 
-        if let Err(e) = cleanup_orphans(&mut *db, self.config.orphan_storage_capacity) {
-            warn!(target: LOG_TARGET, "Failed to clean up orphans: {}", e);
+            if metadata.is_pruned_node() {
```

Comment:
I prefer keeping the logic in a function, its easier to read

Reply:
Reverting back to the previous version, logic is unchanged